src/coreclr/src/vm/jitinterface.cpp

```diff
@@ -12425,7 +12425,13 @@ CorJitResult CallCompileMethodWithSEHWrapper(EEJitManager *jitMgr,
 #if !defined(TARGET_X86)
     if (ftn->HasUnmanagedCallersOnlyAttribute())
     {
-        COMDelegate::ThrowIfInvalidUnmanagedCallersOnlyUsage(ftn);
+        // If the stub was generated by the runtime, don't validate
+        // it for UnmanagedCallersOnlyAttribute usage. There are cases
+        // where the validation doesn't handle all of the cases we can
+        // permit during stub generation (e.g. Vector2 returns).
+        if (!ftn->IsILStub())
```

Comment:
Could you elaborate why this is a problem today? I'm not quite connecting how or why we have special handling for `Vector2/3/4`?

Reply:
It was always a problem given our the existing checking mechanism. We recently moved this check to catch _all_ case instead of just cases that result from a query by the JIT for a `getCallInfo()`. Now we don't miss any and this deficiency in our validation was discovered. The `Vector*` types were just the thing that popped,

Reply:
MarshalingRequired rejects non-primitive value types here: https://github.com/dotnet/runtime/blob/ddd458e8beb5ca7711b6082aeb8ed88f6580512e/src/coreclr/src/vm/dllimport.cpp#L3506

The problem is that the JIT did not always get the return buffers right for non-primitive value types and so the interop IL stub generation compensated for it via a very complex hack.

This should go away once https://github.com/dotnet/runtime/issues/12375 is fixed.

Reply:
I have opened https://github.com/dotnet/runtime/issues/35928 so that we make the unmanaged pointers returning non-primitive value types work for .NET 5 at least.